### PR TITLE
Turn globals.argv0 into a DArray

### PR DESCRIPTION
**Pull request by Geod24**
This is a step towards converting `globals` to use D array.

There is currently no way to call a function accepting a D string from C++ (thanks to ABI issues on i386 / Win64) but it does not matter for `globals` AFAICS.
So I resorted to providing trivial C++ wrapper to the D functions, and gradually removing the wrapper as they become unused, as can be seen in filename.

### CODE REVIEW

File: src/dmd/mars.d [hunk 1]
```diff
@@ -1418,7 +1415,7 @@ private void printPredefinedVersions(FILE* stream)
 
 extern(C) void printGlobalConfigs(FILE* stream)
 {
-    stream.fprintf("binary    %s\n", global.params.argv0);
+    stream.fprintf("binary    %*s\n", global.params.argv0.length, global.params.argv0.ptr);
```
Comment (ibuclaw) [inline src/dmd/mars.d]: `%.*s` - if memory serves me right.

**Review by ibuclaw — COMMENTED** [review]
Is github loosing my reviews?

**Comment by ibuclaw** [comment]
> Is github loosing my reviews?
Answer, yes. This is twice now I've written something.

**Comment by JinShil** [comment]
> Is github loosing my reviews?

I experienced that yesterday as well; started questioning my sanity.

File: src/dmd/frontend.d [hunk 2]
```diff
@@ -79,7 +79,7 @@ string findDMDConfig(string dmdFilePath)
     import dmd.dinifile : findConfFile;
     import std.string : fromStringz, toStringz;
 
-    auto f = findConfFile(dmdFilePath.toStringz, "dmd.conf");
+    auto f = findConfFile(dmdFilePath.toStringz[0 .. dmdFilePath.length], "dmd.conf");
```
Comment (jacob-carlborg) [inline src/dmd/frontend.d]: This is converted to a C string then back to a D string. Am I missing something?

Comment (Geod24) [reply src/dmd/frontend.d]: It was a leftover. `findConfFile` was still dependent on `\0` termination, as one of the function was not converted to `const(char)[]`. I converted it and now `\0` termination isn't required anymore.

**Comment by WalterBright** [comment]
The subject has nothing to do with the code. globals.argv0 is never mentioned in any of the diffs.

File: src/dmd/dmodule.d [hunk 3]
```diff
@@ -943,7 +943,7 @@ extern (C++) final class Module : Package
             assert(prev);
             if (Module mprev = prev.isModule())
             {
-                if (FileName.compare(srcname, mprev.srcfile.toChars()) != 0)
+                if (!FileName.equals(srcname, mprev.srcfile.toChars()))
```
Comment (WalterBright) [inline src/dmd/dmodule.d]: I don't understand. What is being gained here?

File: src/dmd/root/filename.d [hunk 4]
```diff
@@ -49,30 +50,25 @@ nothrow:
...
-    {
         version (Windows)
         {
-            return stricmp(name1, name2);
+            return _strnicmp(name1.ptr, name2.ptr, name1.length) == 0;
         }
         else
         {
-            return strcmp(name1, name2);
+            return strncmp(name1.ptr, name2.ptr, name1.length) == 0;
```
Comment (WalterBright) [inline src/dmd/root/filename.d]: strncmp is the wrong answer here. I quote the C Standard: "The strncmp function compares not more than n characters (characters that follow a null character are not compared) from the array pointed to by s1 to the array pointed to bys 2." I.e. embedded 0's are still a factor. Use memcmp() and memicmp() instead. I've never found a legitimate use for strncmp().

File: src/dmd/root/filename.d [hunk 5]
```diff
@@ -49,30 +50,25 @@ nothrow:
...
-    extern (C++) bool equals(const RootObject obj) const pure
-    {
-        return compare(obj) == 0;
-    }
-
+    /// Compare two name according to the platform's rules (case sensitive or not)
     extern (C++) static bool equals(const(char)* name1, const(char)* name2) pure
     {
-        return compare(name1, name2) == 0;
+        return equals(name1.toDString, name2.toDString);
```
Comment (WalterBright) [inline src/dmd/root/filename.d]: Please use ( ) when calling functions.

File: src/dmd/utils.d [hunk 6]
```diff
@@ -138,3 +139,44 @@ const(char)[] toDString (const(char)* s) pure nothrow @nogc
...
+   Returns:
+     The return value of `T`
+*/
+auto toCStringThen(alias dg)(const(char)[] src)
+{
+    char[512] small = void;
+    scope ptr = (src.length < (small.length - 1))
+                    ? small[0 .. src.length + 1]
+                    : (cast(char*)mem.xmalloc(src.length + 1))[0 .. src.length + 1];
+    scope (exit)
```
Comment (WalterBright) [inline src/dmd/utils.d]: Using scope in dmd is problematic - take a look at the code generated for it, it isn't pretty. It's a lot better if the function is marked nothrow.

File: src/dmd/utils.d [hunk 7]
```diff
@@ -138,3 +139,44 @@ const(char)[] toDString (const(char)* s) pure nothrow @nogc
...
+     one has to copy it.
+
+   Params:
+     src = Slice to use to call the C function
+     dg  = Delegate to call afterwards
+
+   Returns:
+     The return value of `T`
+*/
+auto toCStringThen(alias dg)(const(char)[] src)
```
Comment (WalterBright) [inline src/dmd/utils.d]: Why add functions that are never called?

**Comment by Geod24** [comment]
@WalterBright : There is a bug with this PR I cannot reproduce locally. So I started to bisect it on the auto-tester. This is not ready for review, and I'll modify the title accordingly.

**Comment by WalterBright** [comment]
Ok. But the size of this is getting large. Please consider more bite size chunks.

**Comment by Geod24** [comment]
> Ok. But the size of this is getting large. Please consider more bite size chunks.

Fair enough, I've took some of those commits to https://github.com/dlang/dmd/pull/8585

**Comment by Geod24** [comment]
This is working now, but based on #8585

**Comment by Geod24** [comment]
@klickverbot : Isn't `DArray` enough ? E.g. can't you access `globals.argv0.{ptr,length}` ?
I should have added a test to the C++ interop.

**Comment by Geod24** [comment]
@wilzbach Any idea what's up with that linker failure ? I see it's during coverage, but I'm not sure whether those two files should be compiled together....

**Comment by Geod24** [comment]
This finally passes and is ready for review

File: src/dmd/frontend.d [hunk 8]
```diff
@@ -79,11 +79,11 @@ string findDMDConfig(const(char)[] dmdFilePath)
     import dmd.dinifile : findConfFile;
     import std.string : fromStringz, toStringz;
 
-    auto f = findConfFile(dmdFilePath.toStringz, "dmd.conf");
+    auto f = findConfFile(dmdFilePath, "dmd.conf");
     if (f is null)
         return null;
 
-    return f.fromStringz.idup;
+    return f.idup;
```
Comment (jacob-carlborg) [inline src/dmd/frontend.d]: You can `idup` a string that is `null`, which will save you the if statement.

File: src/dmd/link.d [hunk 9]
```diff
@@ -1049,15 +1048,15 @@ version (Windows)
...
                 if (FileName.exists(lldpath))
-                    return lldpath;
+                    return lldpath.ptr;
             }
 
             // search PATH to avoid createProcess preferring "link.exe" from the dmd folder
             Strings* paths = FileName.splitPath(getenv("PATH"));
-            if (auto p = FileName.searchPath(paths, "link.exe", false))
-                return p;
+            if (auto p = FileName.searchPath(paths, "link.exe"[], false))
```
Comment (jacob-carlborg) [inline src/dmd/link.d]: Why is this slice needed?

Comment (Geod24) [reply src/dmd/link.d]: Otherwise the call is ambiguous, since we have a `const(char)*` and a `const(char)[]` overload, and manifest constant convert to both.
In a later PR (because this is already quite big) I'll remove the `const(char)*` overload

File: src/dmd/frontend.d (hunk 8 above)
Comment (Geod24) [reply src/dmd/frontend.d]:
```diff
diff --git a/src/dmd/frontend.d b/src/dmd/frontend.d
index 93c5a8a94..9ae64a51a 100644
--- a/src/dmd/frontend.d
+++ b/src/dmd/frontend.d
@@ -79,11 +79,7 @@ string findDMDConfig(const(char)[] dmdFilePath)
     import dmd.dinifile : findConfFile;
     import std.string : fromStringz, toStringz;

-    auto f = findConfFile(dmdFilePath, "dmd.conf");
-    if (f is null)
-        return null;
-
-    return f.idup;
+    return findConfFile(dmdFilePath, "dmd.conf").dup;
 }

 /**
```

Updated thanks

File: src/dmd/root/filename.d [hunk 10]
```diff
@@ -624,20 +705,34 @@ nothrow:
...
+         1 if it exists and is not a directory
+         2 if it exists and is a directory
+     */
     extern (C++) static int exists(const(char)* name)
+    {
+        return exists(name.toDString);
+    }
+
+    /// Ditto
+    extern (D) static int exists(const(char)[] name)
```
Comment (jacob-carlborg) [inline src/dmd/root/filename.d]: Should probably return an enum instead of an int, but that might be something for a separate PR.

File: src/dmd/root/filename.d [hunk 11]
```diff
@@ -938,12 +1053,12 @@ version(Posix)
         fileName = The file name to be converted to an absolute path
     Returns: Whatever `F` returns.
     */
-    auto absPathThen(alias F)(const(char)* fileName)
+    auto absPathThen(alias F)(const(char)[] fileName)
     {
         import core.sys.posix.stdlib: realpath, free;
-        auto absPath = realpath(fileName, null /* realpath allocates */);
+        char* absPath = fileName.toCStringThen!((fn) => realpath(&fn[0], null /* realpath allocates */));
```
Comment (jacob-carlborg) [inline src/dmd/root/filename.d]: `const`?

File: src/dmd/link.d (hunk 9 above)
Comment (jacob-carlborg) [reply src/dmd/link.d]: Ah, ok.

File: src/dmd/root/filename.d (hunk 11 above)
Comment (jacob-carlborg) [reply src/dmd/root/filename.d]: Ah, my bad.

**Review by jacob-carlborg — APPROVED** [review]

**Review by RazvanN7 — APPROVED** [review]

**Comment by BBasile** [comment]
This causes a regression : https://issues.dlang.org/show_bug.cgi?id=19510.

**Comment by Geod24** [comment]
I'll look at it

**Comment by BBasile** [comment]
BTW, sorry since i discovered that your not responsible for the regression. It's more a Windows API call that was not used correctly... the problem was only revealed by your changes.
